src/neurostatslib/solver.py

```diff
@@ -0,0 +1,593 @@
+from typing import Callable, List, Optional, Tuple, Union
+
+import jax.numpy as jnp
+import jaxopt
+from numpy.typing import NDArray
+
+from .base_class import _Base
+from .proximal_operator import prox_group_lasso
+
+__all__ = ["UnRegularizedSolver", "RidgeSolver", "LassoSolver", "GroupLassoSolver"]
+
+
+def __dir__() -> list[str]:
+    return __all__
+
+
+class Solver(_Base, abc.ABC):
+    """
+    Abstract base class for optimization solvers.
+
+    This class is designed to provide a consistent interface for optimization solvers,
+    enabling users to easily switch between different solvers and ensure compatibility
+    with various loss functions and regularization schemes.
+
+    Attributes
+    ----------
+    allowed_optimizers :
+        List of optimizer names that are allowed for use with this solver.
+    solver_name :
+        Name of the solver being used.
+    solver_kwargs :
+        Additional keyword arguments to be passed to the solver during instantiation.
+
+    Methods
+    -------
+    instantiate_solver(loss) :
+        Abstract method to instantiate a solver with a given loss function.
+    get_runner(solver_kwargs, run_kwargs) :
+        Get the solver runner with provided arguments.
+    """
+
+    allowed_optimizers: List[str] = []
+
+    def __init__(
+        self,
+        solver_name: str,
+        solver_kwargs: Optional[dict] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self._check_solver(solver_name)
+        self._solver_name = solver_name
+        if solver_kwargs is None:
+            self._solver_kwargs = dict()
+        else:
+            self._solver_kwargs = solver_kwargs
+        self._check_solver_kwargs(self.solver_name, self.solver_kwargs)
+
+    @property
+    def solver_name(self):
+        return self._solver_name
+
+    @solver_name.setter
+    def solver_name(self, solver_name: str):
+        self._check_solver(solver_name)
+        self._solver_name = solver_name
+
+    @property
+    def solver_kwargs(self):
+        return self._solver_kwargs
+
+    @solver_kwargs.setter
+    def solver_kwargs(self, solver_kwargs: dict):
+        self._check_solver_kwargs(self.solver_name, solver_kwargs)
+        return self._solver_kwargs
+
+    def _check_solver(self, solver_name: str):
+        """
+        Ensure the provided solver name is allowed.
+
+        Parameters
+        ----------
+        solver_name :
+            Name of the solver to be checked.
+
+        Raises
+        ------
+        ValueError
+            If the provided solver name is not in the list of allowed optimizers.
+        """
+        if solver_name not in self.allowed_optimizers:
+            raise ValueError(
+                f"Solver `{solver_name}` not allowed for "
+                f"{self.__class__} regularization. "
+                f"Allowed solvers are {self.allowed_optimizers}."
+            )
+
+    @staticmethod
+    def _check_solver_kwargs(solver_name, solver_kwargs):
```

Comment:
do we need this? or can we just let the solver raise its own exception?

Reply:
i like the fact that the error are raised at initialization and not when the function is called